### PR TITLE
Animate the canvas transitions

### PR DESCRIPTION
**Pull request by youknowriad**
This PR adds some animations to the editor canvas when swithching to template mode back and forth and also when switching the device preview.

### CODE REVIEW

File: packages/edit-post/src/components/visual-editor/style.scss [hunk 1]
```diff
@@ -1,5 +1,8 @@
 .edit-post-visual-editor {
 	position: relative;
+	transition: padding 0.2s linear;
```
Comment (mtias) [inline packages/edit-post/src/components/visual-editor/style.scss]: It'd be nice to try framer here

**Comment by jameskoster** [comment]
Here's what I see:


https://user-images.githubusercontent.com/846565/114565169-eeb05f00-9c68-11eb-93b8-5717bc393a4b.mp4

The animation from Template > Content works quite well. But I find the template flashing in to view when moving from Content > Template a little jarring. To the point where it dilutes the overall effect of the animation. I wonder if we could fade the blocks out / in accordingly to smoothen that a bit?

We will also need to consider how the inspector behaves when we switch to the Template tab. Quick mockup:


https://user-images.githubusercontent.com/846565/114565426-28816580-9c69-11eb-86cb-b077106e83ed.mp4

File: packages/edit-post/src/components/visual-editor/style.scss (hunk 1 above)
Comment (youknowriad) [reply packages/edit-post/src/components/visual-editor/style.scss]: Try now :)

**Comment by youknowriad** [comment]
I refactored the animations to use framer motion and also added a fadein/fadeout transition. It's a bit wild :P (the fading required an extra div)

**Comment by jameskoster** [comment]
I had to `npm install` to get this working, but it's very cool. Do I detect some swing on the animations? :D 

Couple of issues:

1. The initial template load is a bit janky
2. In some situations the edge of the canvas appears at the bottom of the document in content editing


https://user-images.githubusercontent.com/846565/114603396-93de2e00-9c8f-11eb-8845-3c20d1656e85.mp4

I wonder if we need the snackbar now... the animation indicates a significant shift in context.

**Comment by jasmussen** [comment]
I see this:

this

Looks and feels great. 👍 👍

A few things:

- To echo Jay as I so often do, the snackbar feels less necessary here. I'm not profoundly opposed to it, the stacking of notices is mostly due to me testing this out. But if we aren't going to remove it, we should consider rephrasing it to be much simpler. Maybe just "Editing template". 
- Notice that if I set focus in the title block, then click "Edit" to edit the template, now the block tab is becomes activated when I enter template mode, even though no block is selected.
- Is the added div going to be a headache in the future or is it a non-issue?


Things that we should look at separately:

- I'm going to take a look at why the is-link style is not using admin theme colors.
- Looking at this, it's increasingly becoming clear that we need a different pattern for the "Back" button. It points left, directionally, which isn't really the place you came from. I suspect this is best solve by rethinking entirely both how you enter, and how you exit, this template editing mode. 
- In Jay's mockups, as is always always the case, the light gray canvas color looked great. Seeing it in action, I can understand Matías suggestion that we might need something darker to better differentiate.

**Comment by jasmussen** [comment]
Is-link patch here: https://github.com/WordPress/gutenberg/pull/30823

**Comment by youknowriad** [comment]
> The initial template load is a bit janky

yes, that is due to the fact that the dependencies of some of these blocks are not there (in the cache) initially, so there's some jumps as the blocks load their content.

> In some situations the edge of the canvas appears at the bottom of the document in content editing

I'll take a look at this.

I'll  remove the snackbar as well.

**Comment by mtias** [comment]
I don't mind the snackbar for now until we figure out all the other details (how the header transforms, the sidebar, etc).

**Comment by youknowriad** [comment]
> In some situations the edge of the canvas appears at the bottom of the document in content editing

This should be fixed now.

**Comment by youknowriad** [comment]
> Notice that if I set focus in the title block, then click "Edit" to edit the template, now the block tab is becomes activated when I enter template mode, even though no block is selected.

I'm not able to reproduce this 

> Is the added div going to be a headache in the future or is it a non-issue?

I don't think so but you never know 😬

**Comment by youknowriad** [comment]
So we get this in? It's an extra 30kb (react-framer-motion) in edit-post script.

**Review by jasmussen — APPROVED** [review]
Happy to try this. works well for me.

**Comment by youknowriad** [comment]
Looks like this broke some e2e tests, I suspect the extra div.

**Comment by mtias** [comment]
Would you rather merge the CSS transition first while we use framer more extensively for the different viewports and extra transitions?

**Comment by jasmussen** [comment]
If the extra div is mostly for the fade, we could do without that for now.

**Comment by youknowriad** [comment]
The fadding out/in is not possible with CSS transitions because there's no "exit" animations. I think I'm personally fine with the extra 30kb to support this change and allow potentially other similar animations in the future.

We can revisit the decision after we experience these animations more.

**Comment by mtias** [comment]
Let's get it in then :)

File: packages/edit-post/src/components/visual-editor/index.js [hunk 2]
```diff
@@ -109,37 +127,46 @@ export default function VisualEditor( { styles } ) {
...
-						}
-					/>
-				</WritingFlow>
-			</div>
+				<AnimatePresence>
+					<motion.div
+						key={ isTemplateMode ? 'template' : 'post' }
+						initial={ { opacity: 0 } }
+						animate={ { opacity: 1 } }
+					>
```
Comment (ellatrix) [inline packages/edit-post/src/components/visual-editor/index.js:141]: Why the need for a second div? Can't this be part of the `motion.div` above?

Comment (youknowriad) [reply packages/edit-post/src/components/visual-editor/index.js:141]: It needs to be unmounted/remounted to trigger the exit transition and it needs to fade/in out just the content and not the frame.

Comment (ellatrix) [reply packages/edit-post/src/components/visual-editor/index.js:141]: What's the difference visually? Fading out the frame or content would look the same?

Comment (youknowriad) [reply packages/edit-post/src/components/visual-editor/index.js:141]: The frame resizes itself at the same time, if we remount it, we kill that animation.

**Pull request merged**